src/backend/binaryen/wasm_expr_gen.ts

```diff
@@ -1923,108 +1942,176 @@ export class WASMExpressionGen {
-            ? undefined
-            : this.wasmExprGen(targetValue);
-        const memberValueType = member.hasSetter
+        const propType = member.hasSetter
             ? (member.setter as VarValue).type
             : member.valueType;
-        res = this.getOriObjInfoByFindIdx(
+
+        const targetValueRef = this.wasmExprGen(targetValue);
+
+        /* TODO: workaround: quick path may fail, since cast failure */
+        const infcDescTypeRef = this.wasmTypeGen.getWASMObjOriType(infcType);
+        const castedObjRef = binaryenCAPI._BinaryenRefCast(
+            this.module.ptr,
             thisRef,
-            infcType,
-            member,
-            memberValueType,
-            fieldIdx,
-            isSet,
-            callMethod,
-            targetValueRef,
+            infcDescTypeRef,
         );
-        if (member.hasSetter) {
-            res = this.callFuncRef(
-                memberValueType,
-                res,
-                [targetValue],
-                thisRef,
+        const ifEqualTypeId = FunctionalFuncs.isTypeIdEqual(
+            this.module,
+            infcType.typeId,
+            metaRef,
+        );
+        let ifEqualTrue: binaryen.ExpressionRef;
+        if (propType.kind === ValueTypeKind.FUNCTION) {
+            /* if property's value type is function, and typeid is equal, then we can get property from vtable */
+            ifEqualTrue = this.setObjMethod(
+                castedObjRef,
+                valueIdx,
+                infcDescTypeRef,
+                targetValueRef,
+            );
+        } else {
+            /* if property's value type is not function, then it must be a field */
+            ifEqualTrue = this.setObjField(
+                castedObjRef,
+                valueIdx,
+                targetValueRef,
             );
+            // TODO: box & unbox depend on field_type_id
         }
-        return res;
+        const ifEqualFalse = this.dynSetInfcProperty(
+            thisRef,
+            indexRef,
+            flagRef,
+            propType,
+            member.isOptional,
+            propTypeIdRef,
+            targetValueRef,
+        );
+        /* set property from interface */
+        return this.module.if(ifEqualTypeId, ifEqualTrue, ifEqualFalse);
     }
 
-    private setObjFieldWithSetter(
+    private setObjMember(
         member: MemberDescription,
         thisRef: binaryen.ExpressionRef,
         thisTypeRef: binaryen.Type,
-        fieldIdx: number,
+        valueIdx: number,
         targetValue: SemanticsValue,
     ) {
         let res: binaryen.ExpressionRef;
-        let targetFuncRef: binaryen.ExpressionRef;
         const targetValueRef = this.wasmExprGen(targetValue);
-        if (!member.hasSetter) {
-            res = this.setObjField(thisRef, fieldIdx, targetValueRef);
+        if (member.type === MemberType.FIELD) {
+            res = this.setObjField(thisRef, valueIdx, targetValueRef);
```

Comment:
What's the difference between `setObjMember` and `setObjField`?

Reply:
Yes, member includes field and method, maybe we rename `setObjMember` to `setObjProperty` is better.